### ISSUE TITLE
[DNEP 2] Narrow the Project Scope to Deep Learning Fundamentals and LLMs

### ISSUE THREAD

**Issue by jshn9515** [jshn9515/deep-learning-notes#45]
## Motivation

The `deep-learning-notes` project has gradually expanded beyond its original scope and now includes or plans to include multiple largely independent areas, such as computer vision, generative models, multimodal learning, and large language models.

While these topics are all related to deep learning, continuing to develop them in parallel would make the project increasingly broad and difficult to maintain. The project should instead focus on a more coherent learning path centered around deep learning fundamentals and modern large language models.

The main reasons are:

1. **The project is becoming too broad**

   Deep learning has developed into many largely independent subfields, including computer vision, generative modeling, multimodal learning, reinforcement learning, and large language models.

   Covering all of these areas in sufficient depth would gradually turn the project into a general deep learning encyclopedia rather than a focused set of learning notes.

   Topics such as object detection, semantic segmentation, diffusion models, and multimodal models can each grow into substantial standalone learning paths.

2. **The current development focus is increasingly centered on LLMs**

   Most recent work in the project has focused on topics such as Transformer architectures, tokenization, GPT, LLM training, distributed training, inference, KV cache management, vLLM, SGLang, and post-training.

   These topics form a coherent progression:

   ```text
   Deep Learning Fundamentals
   → Sequence Modeling
   → Attention
   → Transformer
   → GPT
   → LLM Training
   → LLM Data
   → Scaling
   → Evaluation
   → Post-training
   → Inference
   → Serving
   ```

   Continuing to expand unrelated branches would take time away from developing this main path in greater depth.

3. **Not every important deep learning topic needs to become a complete learning track**

   Some topics are useful for understanding general deep learning concepts without requiring a complete survey of their corresponding field.

   For example, CNNs and ResNet remain useful because they introduce convolution, hierarchical feature extraction, and residual connections. These ideas are broadly useful and also help motivate architectural components used later in Transformers.

   However, extending this into a complete computer vision curriculum would require additional topics such as object detection, semantic segmentation, detection heads, region proposal networks, feature pyramids, and task-specific evaluation metrics.

   These topics are valuable, but they are not necessary for the main LLM-oriented learning path.

4. **Historical context should support the main narrative rather than become a separate branch**

   RNNs, LSTMs, GRUs, and sequence-to-sequence models provide useful context for understanding why attention mechanisms were introduced.

   They naturally lead into:

   ```text
   RNN
   → LSTM / GRU
   → Seq2Seq
   → Bahdanau Attention
   → Self-Attention
   → Transformer
   ```

   These topics therefore strengthen the main narrative and should remain part of the project.

   In contrast, topics such as VAE, DDPM, DiT, object detection, and semantic segmentation introduce additional branches that do not directly support the transition toward modern LLMs.

5. **A narrower scope allows greater technical depth**

   Large language models already involve a large number of tightly connected topics, including:

   - tokenization;
   - Transformer architecture;
   - modern attention variants such as MQA and GQA;
   - KV cache;
   - memory and compute analysis;
   - mixed precision training;
   - distributed training;
   - scaling laws;
   - training data pipelines;
   - evaluation;
   - post-training;
   - quantization;
   - inference optimization;
   - batching and scheduling;
   - serving systems such as vLLM and SGLang.

   Focusing on these topics makes it possible to explain not only model architectures, but also how modern LLMs are trained and served in practice.

6. **A coherent learning path is more useful than broad topic coverage**

   The project should prioritize continuity between chapters.

   A reader should be able to follow one main progression from basic neural networks to modern LLM systems without repeatedly switching between unrelated application domains.

   Reducing the number of side branches makes the structure easier to understand and makes chapter dependencies more explicit.

7. **Reducing scope lowers long-term maintenance cost**

   Every additional learning track requires continued maintenance of:

   - chapter organization;
   - code examples;
   - figures;
   - dependencies;
   - notebooks;
   - English and Chinese versions;
   - PDF builds;
   - internal links;
   - tests and CI.

   Maintaining fewer but deeper topics is more sustainable for a long-term personal learning project.

## Decision

`deep-learning-notes` will narrow its primary scope to **deep learning fundamentals and large language models**.

The main learning path will follow approximately:

```text
Deep Learning Fundamentals
→ CNN and Residual Networks
→ Sequence Modeling
→ RNN / LSTM / GRU
→ Seq2Seq
→ Attention
→ Transformer
→ Modern Attention and KV Cache
→ GPT
→ LLM Training
→ LLM Data
→ Scaling Laws
→ Evaluation
→ Post-training
→ LLM Inference
→ LLM Serving
```

The project will continue to include foundational topics that directly support this progression, including CNNs, ResNet, normalization, embeddings, recurrent neural networks, sequence modeling, and Transformer fundamentals.

Standalone learning tracks that are not necessary for this progression will no longer be part of the main curriculum.

This includes removing or archiving topics such as:

- Vision Transformer as a standalone computer vision branch;
- AutoEncoder and VAE;
- Diffusion Models;
- DiT;
- object detection;
- semantic segmentation;
- other large computer vision or generative modeling branches that do not directly support the LLM-oriented learning path.

Existing material may be preserved in the repository history or an archive rather than being permanently discarded, but it will no longer be maintained as part of the primary learning path.

Future additions should generally satisfy at least one of the following conditions:

- they are part of the core deep learning foundations required by later chapters;
- they help explain the historical development toward Transformer-based models;
- they are directly relevant to modern LLM architecture, training, evaluation, inference, or serving;
- they provide necessary background for understanding practical LLM systems.

This change keeps the project focused, makes the chapter structure more coherent, and allows future work to go deeper into the areas that are most relevant to the project's current direction.
